### ISSUE TITLE
Markdown headings aren't rendered on open

### ISSUE THREAD

**Issue by stootoon** [millejoh/emacs-ipython-notebook#522]
I'm using ein version 20190404.2325 with GNU Emacs 26.1 on Mac OS High Sierra v. 10.13.2. If I create a new heading cell, it's rendered correctly (see attached ..._correct.png). When I save the notebook, the cell is saved correctly as markdown (see notebook contents below) but when I close and reopen the notebook, the markdown cell is displayed as raw markdown, rather than being rendered as h1 (see attached .._incorrect.png). Any idea why this is happening?

![ein_markdown_correct](https://user-images.githubusercontent.com/5781327/57027489-6891de00-6c34-11e9-9994-6d1ad7f9f98a.png)

![ein_markdown_incorrect](https://user-images.githubusercontent.com/5781327/57027498-6fb8ec00-6c34-11e9-91f9-721e306e2170.png)

{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ein.tags": "worksheet-0",
    "slideshow": {
     "slide_type": "-"
    }
   },
   "source": [
    "# Hello"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "name": "Untitled1.ipynb"
 },
 "nbformat": 4,
 "nbformat_minor": 2
}

**Comment by dickmao** [comment]
Starting with 234cf391f96f68ec7a2cd1ef270fea3ae3985771 , ein no longer renders "heading cells" per the [nbformat v4 change](https://nbformat.readthedocs.io/en/latest/format_description.html#markdown-cells).

I'm wondering how you got `h1:` to appear.  Perhaps v20190404.2325 still lets you do that, and the latest code doesn't.

You may like the fancy h1 rendering, but the [nbformat v4 change](https://nbformat.readthedocs.io/en/latest/format_description.html#markdown-cells) and our buggy handling of it #497 compelled me to eliminate it.

**Comment by stootoon** [comment]
Thanks for the reply. Can you suggest an alternate method to visually
emphasize headings? They were very useful for demarcating different
sections of code, but now they're barely visible.

Am Mi., 1. Mai 2019 um 18:05 Uhr schrieb dickmao <notifications@github.com>:

**Comment by dickmao** [comment]
As an emacs person, I like my fonts a fixed size, but people younger than myself like varying font sizes.

You can try enabling polymode:
```
M-x customize-group RET ein
Toggle Ein:Polymode
```
This will "fontify" the heading text to a different color, but the font size alas will remain the same.

**Comment by dickmao** [comment]
I would add that if you know enough about font-lock, you could change the style of headings within `markdown-mode` to make it bigger.  But this is outside the scope of EIN.

**Comment by millejoh** [comment]
Small history lesson. The IPython notebook did originally have distinct cell types for H1, H2, etc headings and ein respected those, even trying to apply a variable font size as the web-based frontend would do. As @dickmao pointed out, IPython/Jupyter discarded the different cell types several revisions back and now just uses straight markup for all non-code cell's.

Being old, curmudgeonly, and unaccustomed to change I tried to keep the spirit of the old ways alive by fudging in support for cell heading types. You can see this when you create a new cell and call `ein:worksheet-change-cell-type`. You'll see options for code (c), hy (h), markdown (m), raw (r), and heading (1-6). Note that internally markdown and heading cells are stored as markdown text, but EIN knows the difference and will apply the pretty scaled font for heading cells. There used to be some code that would detect, when downloading a notebook, markdown cells that look like heading cells and do the right thing. Apparently that code was lost during the major (and most welcome) overhaul ein has seen in the past 6 months. I can look at reintroducing it, so long as I find some time and the change doesn't break anything else.

**Comment by dickmao** [comment]
At some point, I would like to broach the possibility of making polymode the default (I tested `IJulia` the other day, and it worked out of the box as opposed to having to hardcode julia knowledge into `ein-multilang.el` as we do with python and R).

Having EIN be "smart" about h1-6 within markdown mode is, in my opinion, overreach if polymode ever becomes the default.  The style in which h1-6 gets rendered should belong with markdown-mode, not EIN.

**Comment by stootoon** [comment]
I actually agree with letting markdown handle the styles of h1-6 rather
than doing it separately in EIN. The only problem is that I've found that
markdown cells are rarely rendered as markdown, even if there are headings
present. They just get rendered as raw text, which is unsatisfactory if I
want to use headings to indicate different sections of code. For an
example, see the image below. If markdown rendering was consistently
activated, I'd be happy with that, and I could change font-lock to suit my
needs, but currently it doesn't seem to be. Perhaps there's some EIN
setting that I need to configure to get markdown to work right?

[image: image.png]

Am Mi., 1. Mai 2019 um 20:50 Uhr schrieb dickmao <notifications@github.com>:

**Comment by dickmao** [comment]
> Perhaps there's some EIN
setting that I need to configure to get markdown to work right?

First turn on polymode within EIN by
```
M-x customize-group RET ein
Toggle Ein:Polymode
```
Then toggle-on header scaling by
```
M-x customize-group RET markdown-faces
Markdown Header Scaling
```
![Screenshot_2019-05-02_05-31-14](https://user-images.githubusercontent.com/7578770/57067015-d58c8e80-6c9b-11e9-9efc-bd811085933e.png)

**Comment by stootoon** [comment]
Just tried this (after switching ein versions to the melpa 0.15 stable).

I can't turn on polymode, because the Ein:polymode field doesn't exist when
I do M-x customize-group RET ein.

The second step (turning on Markdown Header Scaling) does seem to work
however, because in standard markdown files the headers now have large font
sizes. But within Ein I still get raw text:

[image: image.png]
Was polymode removed in this version?

Am Do., 2. Mai 2019 um 10:34 Uhr schrieb dickmao <notifications@github.com>:

**Comment by dickmao** [comment]
[Caveat MELPA stable](https://github.com/melpa/melpa/blame/88c44fe81f36144c697c93ef0f0ee7577ef6fe58/README.md#L91-L93y)

From emacs, (same information [here](https://melpa.org/#/getting-started)),
```
M-x package-refresh-contents
M-x package-list-packages
```
Find and install `ein` from the package listing buffer.

**Comment by millejoh** [comment]
I agree that polymode is the future. EIN's attempt to provide multi-mode support has always been hack-y. I only get a little grumpy because polymode breaks my evil/spacemacs keybindings, which is definitely not an insurmountable problem by any means.

@dickmao -  When you feel polymode support is stable enough we can go forward with making it the default setting for ein.__

**Comment by stootoon** [comment]
I initially tried getting the latest version from melpa (20190430.1340).

I had even bigger problems with this one: When creating a new notebook, it
would

1) Hang,
2) Then message that "Untitled.ipynb" is ready. At this point the notebook
file would indeed be created.
3) Normally ein would then load Untitled.ipynb. In this version of ein, not
only is the notebook NOT loaded, but when I click on it, it says "Notebook
Untitled.ipynb pending  open! Retry?"

Pressing either 'y' or 'n' does nothing.

At this point I think my best option is to revert back to the version from
about a couple of months ago that was working.

Am Do., 2. Mai 2019 um 11:23 Uhr schrieb dickmao <notifications@github.com>:

**Comment by stootoon** [comment]
Just wanted to say that after a fresh reinstall of Emacs I was able to get
the latest version of EIN (20190430) working properly. Enabling polymode
and setting the markdown header scalings as you suggested is now working
properly for me. Thanks for your help, and for EIN!

Am Do., 2. Mai 2019 um 11:23 Uhr schrieb dickmao <notifications@github.com>:

**Comment by jsigman** [comment]
This is a great thread and answers many questions. Is it possible now to use `markdown-mode` for rendering of markdown cells?

**Comment by dickmao** [comment]
> This is a great thread

Is it?  For one thing, it's out of date.  Polymode is now on by
default since February.  Markdown cells use `ein:markdown-mode`, a crippled
version of the canonical but unmaintained `markdown-mode`.

As all things TUI, emacs's rendering of web 2.0 "isms" like markdown will not
inspire.
